README.md
# ui-router-metatags
Metatags support for the AngularUI Router

Heavily inspired by [tf-metatags](https://github.com/thiagofesta/tf-metatags) and [angular-ui-router-title](https://github.com/nonplus/angular-ui-router-title).

Google updated they crawler to execute javascript, so if you use this component then your pages will be indexed by google. However, other crawlers - search engines or social services like facebook - still need to see the page after being rendered by javascript. Therefor it is still recommended to use a service like prerender.io

## Installation
```sh
$ npm install ui-router-metatags
or
$ bower install ui-router-metatags
```

## Documentation
**See the sample app for a complete example, otherwise, read below**

Include the script in your index file

```html
<script src="node_modules/ui-router-metatags/dist/ui-router-metatags.min.js"></script>
```

Include it in your module declaration and in your run block

```javascript
angular.module('myApp', ['ui.router', 'ui.router.metatags']);

function runBlock($rootScope, MetaTags) {
	$rootScope.MetaTags = MetaTags;
}

angular
	.module('myApp')
	.run(['$rootScope', 'MetaTags', runBlock]);
```

Add the tags to your index file

```html
<title ng-bind="MetaTags.title">Default title</title>
<meta name="description" content="{{MetaTags.description}}">
<meta name="keywords" content="{{MetaTags.keywords}}">
<meta ng-repeat="(key, value) in MetaTags.properties" property="{{key}}" content="{{value}}" >
<meta name="prerender-status-code" content="{{MetaTags.prerender.statusCode}}">
<meta name="prerender-header" ng-if="MetaTags.prerender.header" content="{{MetaTags.prerender.header}}">
```

Then configure defaults

```javascript
function configure(UIRouterMetatagsProvider) {
    UIRouterMetatagsProvider
        .setTitlePrefix('prefix - ')
        .setTitleSuffix(' | MyApp')
        .setDefaultTitle('MyApp')
        .setDefaultDescription('description')
        .setDefaultKeywords('keywords')
        .setStaticProperties({
                'fb:app_id': 'your fb app id',
                'og:site_name': 'your site name'
            })
        .setOGURL(true);
}

angular
    .module('myApp')
    .config(['UIRouterMetatagsProvider', configure]);
```

(Static properties are added to all pages and the "setOGURL" method ensures that a 'og:url' property is added to all pages.)

And finally decorate the routes with metatags in the route configs like so:

```javascript
function configureRoutes($stateProvider) {
    $stateProvider
        .state('frontpage', {
            url: '/',
            metaTags: {
                title: 'Frontpage',
                description: 'This is the frontpage',
                keywords: 'lots of interresting keywords',
                properties: {
                    'og:title': 'Frontpage'
                }
            }
        })
        .state('blogposts', {
            url: '/blog/:category',
            resolve: {
                /* @ngInject */
                posts: function(myService, $stateParams) {
                    return myService.getPosts($stateParams.category);
                }
            },
            metaTags: {
                prerender: {
                    /* @ngInject */
                    statusCode: function(posts) {
                        return posts.length > 0 ? 200 : 302;
                    },
                    /* @ngInject */
                    header: function(posts) {
                        return posts.length > 0 ? null : 'Location: http://example.com/posts';
                    }
                }
            }
        })
        .state('blogpost', {
            url: '/post/:id',
            resolve: {
                /* @ngInject */
                blogpost: function(myService, $stateParams) {
                    return myService.getPost($stateParams.id);
                }
            },
            metaTags: {
                /* @ngInject */
                title: function(blogpost) {
                    return blogpost.title;
                },
                description: 'The most interresting post {{blogpost.title}}'
            }
        });

}
angular
    .module('myApp')
    .config(['$stateProvider', configureRoutes]);
```

Note that all tags can be either a simple string, a resolve function or a interpolated string (where the properties available are the ones you resolve in your route).


## Log statement
Please note that any routes without metatags will cause a debug log statement, so remember to disable debug logging in production. Example of such a log statement:

```javascript
MetaTags - route: "app.dashboard" does not contain any metatags
```

## To develop
Install nodejs and the following packages globally:
* gulp
* typings

then run:
* npm install
* typings install

Then finally, run "gulp"


CHANGELOG.md
# 1.0.4
Merged https://github.com/tinusn/ui-router-metatags/pull/27
Upgraded build dependencies

# 1.0.3
* ...today is a looooong day.... published the actualy changes made in 1.0.1--

# 1.0.2
* removed debug log

# 1.0.1
* Fixed a bug when using numbers in the prerender.statusCode

# 1.0.0
* Fixes [#2](https://github.com/tinusn/ui-router-metatags/issues/2)
* Added a sample
Breaking changes:
* (Typescript only) Renamed: uiroutermetatags.IService to uiroutermetatags.IConfig

# 0.0.6
Fixes [#1](https://github.com/tinusn/ui-router-metatags/issues/1)
Added to bower

# 0.0.5
Updated dependencies and added the package to [npm](https://www.npmjs.com/package/ui-router-metatags)

# 0.0.4
Added prerender status code and header.

When an stateChangeError or a stateNotFound event is thrown, the status is set automatically (to 500 and 404 respectively), but it also offers manual control if needed.

```javascript
function configureRoutes($stateProvider) {
    $stateProvider
        .state('blogposts', {
            url: '/blog/:category',
            resolve: {
                /* @ngInject */
                posts: function(myService, $stateParams) {
                    return myService.getPosts($stateParams.category);
                }
            }
            metaTags: {
                prerender: {
                    /* @ngInject */
                    statusCode: function(posts) {
                        return posts.length > 0 ? 200 : 302;
                    },
                    /* @ngInject */
                    header: function(posts) {
                        return posts.length > 0 ? null : 'Location: http://example.com/posts';
                    }
                }
            }
        });
}
angular
    .module('myApp')
    .config(configureRoutes);
```

In the example above, if there are posts, then a status code of 200 is returned, whereas if no posts are there, then a 302 is set and a redirect location is offered.


# 0.0.3
Added the possibility to have static tags and to set the 'og:url' property by default.

```javascript
function configure(UIRouterMetatagsProvider) {
    UIRouterMetatagsProvider
        .setTitlePrefix('prefix - ')
        .setTitleSuffix(' | MyApp');
        .setDefaultTitle('MyApp')
        .setDefaultDescription('description')
        .setDefaultKeywords('keywords')
        .setStaticProperties({
                'fb:app_id': 'your fb app id',
                'og:site_name': 'your site name'
            })
        .setOGURL(true);
}

angular
    .module('myApp')
    .config(configure);
```

# 0.0.2
Tags can now resolve properties resolved by ui-router

Example:

```javascript
function configureRoutes($stateProvider) {
    $stateProvider
        .state('blogpost', {
            url: '/post/:id',
            resolve: {
                /* @ngInject */
                blogpost: function(myService, $stateParams) {
                    return myService.getPost($stateParams.id);
                }
            }
            metaTags: {
                /* @ngInject */
                title: function(blogpost) {
                    return blogpost.title;
                },
                description: 'The most interresting post {{blogpost.title}}'
            }
        });


}
angular
    .module('myApp')
    .config(configureRoutes);
```

# 0.0.1
Initial version

gulpfile.js
/* globals require */

'use strict';

var concat = require('gulp-concat');
var gulp = require('gulp');
var ngAnnotate = require('gulp-ng-annotate');
var sourcemaps = require('gulp-sourcemaps');
var ts = require('gulp-typescript');
var uglify = require('gulp-uglify');
var usemin = require('gulp-usemin');
var rimraf = require('rimraf');
var rename = require('gulp-rename');
var merge = require('merge2');

var tsProject = ts.createProject({
    declaration: true,
    noResolve: true,
    noImplicitAny: false,
    removeComments: false,
    target: 'ES5',
    emitDecoratorMetadata: false
});

gulp.task('typescript', ['clean'], function () {
    var tsResult = gulp.src(['src/*.module.ts', 'src/*.ts', 'typings/**/*.d.ts'])
        .pipe(sourcemaps.init())
        .pipe(tsProject());

    var jsResult = tsResult.js
        .pipe(concat('ui-router-metatags.js'))
        .pipe(ngAnnotate({
            gulpWarnings: true
        }))
        .pipe(sourcemaps.write('./'))
        .pipe(gulp.dest('dist/'));

    var dtsResult = tsResult.dts
        .pipe(gulp.dest('dist/'));

    return merge([
        jsResult,
        dtsResult
    ]);
});

gulp.task('usemin', ['typescript'], function () {
    return gulp.src('dist/*.js')
        .pipe(sourcemaps.init({
            loadMaps: true
        }))
        .pipe(uglify())
        .pipe(rename({
            suffix: '.min'
        }))
        .pipe(sourcemaps.write('./'))
        .pipe(gulp.dest('dist'));

});

gulp.task('clean', function (cb) {
    rimraf('dist', cb);
});

gulp.task('default', function () {
    gulp.start('usemin');
});


src/ui-router-metatags.ts
/**
 * Metatags for angular-ui-router
 */
namespace uiroutermetatags {
	const appModule = angular.module('ui.router.metatags', ['ui.router']);

	export interface IMetaTags {
		title?: string | Function;
		description?: string | Function;
		keywords?: string | Function;
		robots?: string | Function;
		properties?: {
			[index: string]: string | Function;
		},
		prerender?: Prerender;
	}

	export interface Prerender {
		statusCode?: number | Function,
		header?: string | Function
	}

	export interface IProvider {
		setTitlePrefix(prefix: string): IProvider;
		setTitleSuffix(suffix: string): IProvider;
		setDefaultTitle(title: string): IProvider;
		setDefaultDescription(description: string): IProvider;
		setDefaultKeywords(keywords: string): IProvider;
		setDefaultRobots(robots: string): IProvider;
		setStaticProperties(properties: {}): IProvider;
		setOGURL(enabled: boolean): IProvider;
	}

	export interface IService {
		update(tags: IMetaTags, transition: any): void;
		prerender: uiroutermetatags.Prerender;
	}

	export interface IConfig {
		prefix: string;
		suffix: string;
		defaultTitle: string;
		defaultDescription: string;
		defaultKeywords: string;
		defaultRobots: string;
		staticProperties: {};
		enableOGURL: boolean;
	}

	class UIRouterMetatags implements angular.IServiceProvider, uiroutermetatags.IProvider {
		prefix: string = '';
		suffix: string = '';
		defaultTitle: string = '';
		defaultDescription: string = '';
		defaultKeywords: string = '';
		defaultRobots: string = '';
		staticProperties: {} = {};
		enableOGURL: boolean = false;

		/* @ngInject */
		constructor() {

		}

		setTitlePrefix(prefix: string): UIRouterMetatags {
			this.prefix = prefix;
			return this;
		}

		setTitleSuffix(suffix: string): UIRouterMetatags {
			this.suffix = suffix;
			return this;
		}

		setDefaultTitle(title: string): UIRouterMetatags {
			this.defaultTitle = title;
			return this;
		}

		setDefaultDescription(description: string): UIRouterMetatags {
			this.defaultDescription = description;
			return this;
		}

		setDefaultKeywords(keywords: string): UIRouterMetatags {
			this.defaultKeywords = keywords;
			return this;
		}

		setDefaultRobots(robots: string): UIRouterMetatags {
			this.defaultRobots = robots;
			return this;
		}

		setStaticProperties(properties: {}): UIRouterMetatags {
			this.staticProperties = properties;
			return this;
		}

		setOGURL(enabled: boolean): UIRouterMetatags {
			this.enableOGURL = enabled;
			return this;
		}

		public $get(): uiroutermetatags.IConfig {
			return {
				prefix: this.prefix,
				suffix: this.suffix,
				defaultTitle: this.defaultTitle,
				defaultDescription: this.defaultDescription,
				defaultKeywords: this.defaultKeywords,
				defaultRobots: this.defaultRobots,
				staticProperties: this.staticProperties,
				enableOGURL: this.enableOGURL
			}
		}
	}

	appModule.provider('UIRouterMetatags', UIRouterMetatags);

	class MetaTags {
		title: string;
		keywords: string;
		description: string;
		robots: string;
		properties: {};
		prerender: uiroutermetatags.Prerender = {};
		
		/* @ngInject */
		constructor(public $log: angular.ILogService, public UIRouterMetatags: uiroutermetatags.IConfig, 
			public $interpolate: angular.IInterpolateService, public $injector: angular.auto.IInjectorService, 
			public $state: any, public $location: angular.ILocationService, public $window,
			public $uiRouterGlobals: any
		) {
		}

		update(tags: uiroutermetatags.IMetaTags, transition: any) {
			try {

				/* 
					This hack is used to resolve the lack of "this.$state.$current.locals.globals".
					Here the resolved object is builded with all previews resolved data
				*/
				var resolved = {}
				let tokens = transition.getResolveTokens();
				tokens.forEach(token => resolved[token] = transition.injector().get(token));

				this.properties = angular.extend({}, this.UIRouterMetatags.staticProperties);

				if (this.UIRouterMetatags.enableOGURL) {
					this.properties['og:url'] = this.$location.absUrl();
				}

				if (tags) {
					this.title = tags.title ? this.UIRouterMetatags.prefix + (this.getValue('title', tags.title, resolved) || '') + this.UIRouterMetatags.suffix : this.UIRouterMetatags.defaultTitle;
					this.description = tags.description ? this.getValue('description', tags.description, resolved) : this.UIRouterMetatags.defaultDescription;
					this.keywords = tags.keywords ? this.getValue('keywords', tags.keywords, resolved) : this.UIRouterMetatags.defaultKeywords;
					this.robots = tags.robots ? this.getValue('robots', tags.robots, resolved) : this.UIRouterMetatags.defaultRobots;
					angular.forEach(tags.properties, (value, key) => {
						var v = this.getValue(key, value, resolved);
						if (v) {
							this.properties[key] = v;
						}
					});
				} else {
					this.title = this.UIRouterMetatags.defaultTitle;
					this.description = this.UIRouterMetatags.defaultDescription;
					this.keywords = this.UIRouterMetatags.defaultKeywords;
					this.robots = this.UIRouterMetatags.defaultRobots;
				}
				if (tags && tags.prerender) {
					this.prerender.statusCode = tags.prerender.statusCode ? this.getValue('prerender.statusCode', tags.prerender.statusCode, resolved) : 200;
					this.prerender.header = tags.prerender.header ? this.getValue('rerender.header', tags.prerender.header, resolved) : null;
				} else {
					this.prerender.statusCode = 200;
					this.prerender.header = null;
				}

				this.$window.prerenderReady = true;
			} catch (err) {
				this.$log.error('error occured when extracting metatags:', err);
			}
		}

		getValue(tagType: string, tag, resolved: any) {
			try {
				if (!tag) {
					return null;
				} else if (typeof tag === 'number') {
					return tag;
				} else if (typeof tag === 'string' && tag.trim().length === 0) {
					return null;
				} else if (angular.isFunction(tag) || Array.isArray(tag)) {
					return this.$injector.invoke(tag, this, resolved);
				} else {
					return this.$interpolate(tag)(resolved);
				}
			} catch (err) {
				this.$log.error('error occured when trying to get the value of tag:', tagType, err);
				return null;
			}
		}
	}

	appModule.service('MetaTags', MetaTags);

	/* @ngInject */
	function runBlockTransitions($log: angular.ILogService, MetaTags: uiroutermetatags.IService, $window: angular.IWindowService, $transitions: any) {

		$transitions.onStart({}, onStart)
		$transitions.onSuccess({}, onSuccess)
		$transitions.onError({}, onError)


		function onStart(transition: any) {
			$window.prerenderReady = false;
		}

		function onSuccess(transition: any) {
			var toState = transition.$to()
			var fromState = transition.$from()
			
			if (!toState.metaTags) {
				$log.debug(`MetaTags - route: "${toState.name}" does not contain any metatags`);
			}

			MetaTags.update(toState.metaTags as uiroutermetatags.IMetaTags, transition);
		}

		function onError(transition: any) {
			MetaTags.prerender.statusCode = 500;
			$window.prerenderReady = true;
		}
	}
	
	appModule.run(runBlockTransitions);
}

declare module angular.ui {
    interface IState {
		metaTags?: uiroutermetatags.IMetaTags;
	}
}

interface Window {
	prerenderReady?: boolean;
}


src/ui-router-metatags.d.ts
declare namespace uiroutermetatags {
	interface IMetaTags {
		title?: string|Function;
		description?: string|Function;
		keywords?: string|Function;
		properties?: {
			[index: string]: string|Function;
		},
		prerender?: Prerender;
	}

	interface Prerender {
		statusCode?: number|Function,
		header?: string|Function
	}

	interface IProvider {
		setTitlePrefix(prefix: string): IProvider;
		setTitleSuffix(suffix: string): IProvider;
		setDefaultTitle(title: string): IProvider;
		setDefaultDescription(description: string): IProvider;
		setDefaultKeywords(keywords: string): IProvider;
		setStaticProperties(properties: {}): IProvider;
		setOGURL(enabled: boolean): IProvider;
	}

	interface IService {
		prefix: string;
		suffix: string;
		defaultTitle: string;
		defaultDescription: string;
		defaultKeywords: string;
		staticProperties: {};
		enableOGURL: boolean;
	}
}

declare module angular.ui {
    interface IState {
		metaTags?: uiroutermetatags.IMetaTags;
	}
}

interface Window {
	prerenderReady?: boolean;
}

sample/app.js
angular.module('myApp', [
	'ui.router',
	'ui.router.metatags',
]);

function configure($urlRouterProvider, $stateProvider, UIRouterMetatagsProvider) {
	UIRouterMetatagsProvider
		.setTitlePrefix('prefix - ')
		.setTitleSuffix(' | MyApp')
		.setDefaultTitle('MyApp')
		.setDefaultDescription('description')
		.setDefaultKeywords('keywords')
		.setDefaultRobots('index')
		.setStaticProperties({
			'fb:app_id': 'your fb app id',
			'og:site_name': 'your site name'
		})
		.setOGURL(true);

	$urlRouterProvider.otherwise('/');

    $stateProvider
		.state('app', {
			abstract: true,
			template: '<h1>angular-ui-router sample</h1><div ui-view></div><h2>Resolved tags in MetaTags</h2><pre>title: {{MetaTags.title | json}}</pre><pre>keywords: {{MetaTags.keywords | json}}</pre><pre>description: {{MetaTags.description | json}}</pre><pre>robots: {{MetaTags.robots | json}}</pre><pre>properties: {{MetaTags.properties | json}}</pre><pre>prerender: {{MetaTags.prerender | json}}</pre>',
		  })
        .state('app.frontpage', {
            url: '/',
			template: "<h1>frontpage</h1><a ui-sref=\"app.posts({category: 'test'})\">blog posts</a>",
            metaTags: {
                title: 'Frontpage',
                description: 'This is the frontpage',
                keywords: 'lots of interresting keywords',
                properties: {
                    'og:title': 'Frontpage'
                }
            }
        })
        .state('app.posts', {
            url: '/blog/:category',
			template: '<h1>blog posts</h1><ul><li ng-repeat="post in posts"><a ui-sref="app.post({id: post.id})">{{post.title}}</a></li></ul><a ui-sref="app.frontpage">back to the frontpage</a>',
			controller: function ($scope, posts) {
				$scope.posts = posts;
			},
            resolve: {
                /* @ngInject */
                posts: function (myService, $stateParams) {
                    return myService.getPosts($stateParams.category);
                }
            },
            metaTags: {
				title: 'Blog posts',
				robots: 'noindex, follow',
                prerender: {
                    /* @ngInject */
                    statusCode: function (posts) {
                        return posts.length > 0 ? 200 : 302;
                    },
                    /* @ngInject */
                    header: function (posts) {
                        return posts.length > 0 ? null : 'Location: http://example.com/posts';
                    }
                }
            }
        })
        .state('app.post', {
            url: '/post/:id',
			template: '<h1>{{post.title}}</h1><p>{{post.description}}</p><a ui-sref="app.frontpage">back to the frontpage</a>',
			controller: function ($scope, post) {
				$scope.post = post;
			},
            resolve: {
                /* @ngInject */
                post: function (myService, $stateParams) {
                    return myService.getPost($stateParams.id);
                }
            },
            metaTags: {
                title: function (post) {
                    return post.title;
                },
				robots: 'nofollow',
                description: 'The most interresting post {{post.title}}'
            }
        });
}

angular
	.module('myApp')
	.config(['$urlRouterProvider', '$stateProvider', 'UIRouterMetatagsProvider', configure]);

function runBlock($rootScope, MetaTags) {
	$rootScope.MetaTags = MetaTags;
}

angular
	.module('myApp')
	.run(['$rootScope', 'MetaTags', runBlock]);

function myService() {
	var service = {
		getPosts: getPosts,
		getPost: getPost
	};

	var samplePost = {
		id: 1,
		title: 'sample blog post',
		content: 'lorem ipsum'
	};

	function getPosts(category) {
		return [samplePost];
	}

	function getPost(id) {
		return samplePost;
	}

	return service;
}

angular
	.module('myApp')
	.service('myService', myService);




